### PR TITLE
feat(editors/subscriber/later-binding): add input requirement check

### LINKED ISSUE #867: Subscriber Later Binding: Check for CDC and bType
As a user of OpenSCD I want to be warned that a later binding type `ExtRef` does not meet the requirements based on the attributes `pLN`, `pDO` and `pDA`.

**Requirements:**
- Check if `cdc` of both `ExtRef` and the `FCDA` does meet
- Check if `bType` of both `ExtRef` and `FCDA` meet

**Additional Information: **
Getting `CDC` for `FCDA`:
1. Take `ldInst`, `prefix`, `lnClass` and `lnInst`
2. Navigate to the `LN0` or `LN` in the source IED (the IED where `FCDA` resides in)
3. Take `lnType` attribute
4. Navigate to the `LNodeType` (`lnType` equals the `id` attribute of `LNodeType`)
5. Take the `doName` from the `FCDA`
6. Navigate to the `DO` element
7. Take the `type` attribute
8. Navigate to the `DOType` (`type` equals `id` attribute of the `DOType`)
9. Take the `cdc` attribute -> Done!

> In case `doName` has more than one name `A.phsA` and is seperated by `.` the steps 6 to 8 have to be repeated!

Getting CDC of `ExtRef`:
1. Take `pLN`
2. Navigate to the `LN0` or `LN` in the sink IED (the IED where `ExtRef` resides in) 
3. Do steps 3 to 9 (in step 5 take `pDO` instead of `doName`)
> In case `doName` has more than one name `A.phsA` and is seperated by `.` the steps 6 to 8 have to be repeated!

Getting `bType` of `FCDA`:
1. do steps 1 to 8
2. Take `daName` from `FCDA`
3. Navigate to `DA`/`BDA` element
4. Take the `bType` element
> In case `daName` has more than one name `cVal.mag.f` and is seperated by `.` the `BDA` is a child of a `DAType`.

Getting `bType` of `ExtRef`:
Same as the above, except that `pDA` has to be taken instead of `daName`

### PR DESCRIPTION
**Pull request by JakobVogelsang**
Closes #867 

Check that the selected data and a later binding external reference match in terms of their `CDC` and `bType` specification.

### CODE REVIEW

**Review by danyill — COMMENTED** [review]
Just one real question/comment from me.

File: src/editors/subscription/later-binding/foundation.ts [hunk 1]
```diff
@@ -1,7 +1,104 @@
...
+} {
+  const [pLN, pDO, pDA] = ['pLN', 'pDO', 'pDA'].map(attr =>
+    extRef.getAttribute(attr)
+  );
+  if (!pLN || !pDO || !pDA) return { cdc: null, bType: null };
+
+  const anyLn = extRef
+    .closest('IED')
+    ?.querySelector(`LN[lnClass="${pLN}"],LN0[lnClass="${pLN}"]`);
+  if (!anyLn) return { cdc: null, bType: null };
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply src/editors/subscription/later-binding/foundation.ts:97]: That is very much correct. I propose to query them all and find the first where we can extract `CDC` and `bType`. It still is not perfect but with the loose coupling (missing prefix and ln instance in the requirements of the `ExtRef`) it is the best we can do without insane effort.
Thx @danyill

File: src/editors/subscription/later-binding/foundation.ts [hunk 2]
```diff
@@ -1,7 +1,104 @@
...
+
+  const doNames = doName.split('.');
+  let leaf: Element | null | undefined = lNodeType;
+  for (const doName of doNames) {
+    const dO: Element | null | undefined = leaf?.querySelector(
+      `DO[name="${doName}"], SDO[name="${doName}"]`
+    );
+    leaf = doc.querySelector(`DOType[id="${dO?.getAttribute('type')}"]`);
+  }
+  if (!leaf || !leaf.getAttribute('cdc')) return { cdc: null, bType: null };
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply src/editors/subscription/later-binding/foundation.ts:22]: Thx. We can make a util wishlist :). I am thinking for a long time now about something like this inside `open-scd-core` or even outside of it in an extra repository. I would start a discussion for this here #1051

File: src/editors/subscription/later-binding/ext-ref-later-binding-list.ts [hunk 3]
```diff
@@ -71,19 +75,40 @@ export class ExtRefLaterBindingList extends LitElement {
...
   /**
-   * The data attribute check using attributes pLN, pDO, pDA and pServT is not supported yet by this plugin.
-   * To make sure the user does not do anything prohibited, this type of ExtRef cannot be manipulated for the time being.
-   * (Will be updated in the future).
+   * Check data consistency of source `FCDA` and sink `ExtRef` based on
+   * `ExtRef`'s `pLN`, `pDO`, `pDA` and `pServT` attributes.
+   * Consistent means `CDC` and `bType` of both ExtRef and FCDA is equal.
+   * In case
+   *  - `pLN`, `pDO`, `pDA` or `pServT` attributes are not present allow subscribing
+   *  - no CDC or bType can be extracted allow subscribing (not restrictive on doubt)
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply src/editors/subscription/later-binding/ext-ref-later-binding-list.ts]: Thanks not even spelling incorrect but also just wrong. :D I fixed it

File: src/editors/subscription/later-binding/foundation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply src/editors/subscription/later-binding/foundation.ts]: Fixed thx

**Comment by danyill** [comment]
The changes look good.

I have, I think found an issue in testing some files although it may be unrelated to this new feature and more related to some other later binding issue which I keep seeing but haven't nailed down and I must investigate.

image

I have created a more minimal example:

[XAT_Prot2_Preconfigured_Check (1).scd.zip](https://github.com/openscd/open-scd/files/9877777/XAT_Prot2_Preconfigured_Check.1.scd.zip)

For IED `XAT_232_P2`, there exists an unmapped ExtRef:

```xml
<ExtRef desc="RxExtIn1" intAddr="RxExtIn1;/Ind/stVal" pServT="GOOSE" pDO="Ind" pDA="stVal"/>
<ExtRef desc="RxExtIn1" intAddr="RxExtIn1;/Ind/q" pServT="GOOSE" pDO="Ind" pDA="q"/>
```

Both `stVal` and `q` have preconfigured restrictions (I manually removed the pre-existing mapping).

I expect on the right-hand side that the q values are greyed out because the bType for a general (boolean) and a quality value (Quality) are distinct.

I can look at this in more detail for some time if this is not clear enough -- but perhaps I have misunderstood?

**Comment by JakobVogelsang** [comment]
The `ExtRef` is missing `pLN`. So far we do not restrict so long not all four attributes `pLN`, `pDO`, `pDA` and `pServT` are not given. Do you think this is incorrect @danyill ?

**Review by danyill — APPROVED** [review]

**Comment by danyill** [comment]
> The ExtRef is missing pLN. So far we do not restrict so long not all four attributes pLN, pDO, pDA and pServT are not given. Do you think this is incorrect @danyill ?

Thank you for looking.
This whole referencing system is dubious as you point out.

I think Part 6, 9.3.13 allows for partial `pXX` referencing:

> In this case any values at pDO and possibly pDA allow to specify the CDC
respective attribute type expected for this intAddr. Together with lnClass this allows to specify
a functionally expected input. The desc attribute allows to give it additional informal meaning,
the serviceType attribute the intended service quality. If at least the pDO attribute is specified,
then a missing pDA attribute includes at least all the operational value attribute(s) of the DO,
i.e. stVal, mag, t, q etc. It might allow binding of additional attributes supported by the service
type. 

And also:

> Specifying expected inputs (input templates) of logical devices or logical nodes from
the view of the IED configuration tool. The intAddr attribute shall be set to a non empty
value. The pDO might specify the needed input CDC via a standardized DO name,
additional pDA values might indicate the expected data type of an attribute and the
pLN might document the expected function providing the input. pServT might provide
the expected service type for this input value. These attribute values, if supplied by the
IED configurator, shall not be modified by the system configurator. A system
configurator binding to these input templates must assure that the CDC specified by
pDO and, if given, the attribute (base) type specified by pDA are met. For a final
engineered data flow also the pServT, if given, shall be met by the serviceType value.

This seems to indicate that `pLN` can be optional.

As long as we have a "relaxed" approach where the provision of `pXX` but not finding a match allows subscriptions to be done I think we are all good. We can improve this feature in future on request or as the need arises.

I do see some benefit in being able to specify only `pDO` and `pDA` because it means that the quality items will be nicely greyed out most of the time.

I think we should merge this unless you would like to do more :+1:

**Comment by JakobVogelsang** [comment]
I would merge this as is. It goes easy on the fully specified restriction at the moment. I think this in the future we will see more vendors using all attributes than the other way around. To make the implementation adjust more with the idea in the standard, I would propose another algorithm that goes something like this.

1. get all `DO`, `SDO` based on the `pDO`
2. filter for option `lnClass` but at least filter `DO`, `SDO` used by the subscribing IED
4. map to `DOType`s `cdc`. 
5. filter all null and or undefined
6. put into Set. 
7. take the first entry. There should be only one 

for data attributes
1. get all `DA`, `BDA` based on the `pDA`
2. filter for optional `lnClass` but at least filter `DA`, `BDA` used by the subscribing IED
3. optional filter for given 
4. map to `DA`s and `BDA`s `bType`. 
5. filter all null and or undefined
6. put into Set. 
7. take the first entry. There should be only one 

That could be fast search but also not really reliable. LN to DO to DA is clear, but the other way around is not. However, we want to not be restrictive if we are not sure, so this would even fit the concept.

**Pull request merged**